File: app/Market/h/src/Conf.hs
-- Copyright 12-Dic-2018 ºDeme
-- GNU General Public License - V3 <http://www.gnu.org/licenses/>

-- | Configuration data

module Conf (
  get,
  set
  ) where

import Text.Printf
import qualified Dm.File as File
import qualified Dm.Js as Js
import Dm.Js (JSValue)
import qualified Global as G

path = G.path ["data", "conf.db"]

-- | @'get'@ - Returns configuration data and intializes them if is need.
get :: IO JSValue
get = do
  ex <- File.exists path
  if ex
  then do
    d <- File.read path
    return $ Js.fromStr d
  else
    let d = Js.wMap [("lang", Js.wString "en"),
                     ("menu", Js.wString "settings"),
                     ("bet", Js.wInt 15000),
                     ("allCos", Js.wBool True)]
    in  do
      File.write path (Js.toStr d)
      return d

-- | @'set' key value@ - Modifies a configuration entry.
set :: String -> JSValue -> IO ()
set key value = do
  d <- get
  let dJs = Js.rMap d
  case lookup key dJs of
    Nothing -> error $ printf "Key '%s' not found" key
    Just js ->
      File.write path $ Js.toStr $ Js.wMap $ (key, value):filter f dJs
  where
    f (k, _) = k /= key

File: app/KtMarket/go/data/conf/conf.go
// Copyright 03-Jun-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

//Configuration data.
package conf

import (
	"github.com/dedeme/KtMarket/data/activity"
	"github.com/dedeme/ktlib/js"
)

type T struct {
	Lang     string
	Activity *activity.T
}

// Constructor.
//   lang: 'en' or 'es'.
//   act : Last activity.
func New(lang string, act *activity.T) *T {
	return &T{lang, act}
}

func ToJs(cf *T) string {
	return js.Wa([]string{
		js.Ws(cf.Lang),
		activity.ToJs(cf.Activity),
	})
}

func FromJs(j string) *T {
	a := js.Ra(j)
	return New(
		js.Rs(a[0]),
		activity.FromJs(a[1]),
	)
}
